### PR TITLE
Handle source cluster in error management

### PR DESCRIPTION
**Pull request by rezib**
Commit from @Lorel.

### CODE REVIEW

File: js/core/clusters/clusters.js [hunk 1]
```diff
@@ -85,6 +85,7 @@ define([
         }
 
         config.cluster = cluster;
+        window.cluster = cluster;
```
Comment (rezib) [inline js/core/clusters/clusters.js]: If window.cluster is introduced, config.cluster must be removed from code to avoid duplication.

File: js/utils/ajax.js [hunk 2]
```diff
@@ -19,16 +19,30 @@
...
+  'jquery',
+  'token-utils',
+  'user-utils'
+], function($, token, user) {
   $(document).ajaxError(function(event, jqueryXHR, error, errorThrown) {
+    var i, cluster,
+      isCurrentCluster = window.cluster && error.url.indexOf(window.cluster.api.url) > -1;
+
+    // find concerned cluster
+    for (i in window.clusters) {
```
Comment (rezib) [inline js/utils/ajax.js]: I'm not sure what this loop is supposed to do?

File: js/utils/ajax.js [hunk 3]
```diff
@@ -19,16 +19,30 @@
...
+      cluster = window.clusters[i];
+    }
+
     if (!jqueryXHR.status && !(error.url.indexOf('/authentication') > -1)) {
-      console.log(JSON.stringify(event), JSON.stringify(jqueryXHR), JSON.stringify(error), JSON.stringify(errorThrown));  // eslint-disable-line no-console
-      $('#flash .alert').append($('<p>').text('Error : unable to perform authentication'));
-      $('#flash').show();
+      // logout for concerned cluster
+      if (cluster.api.url === error.url.split('/').slice(0, 3).join('/')) {
```
Comment (rezib) [inline js/utils/ajax.js]: Please explain in comment what does the split+splice+join is trying to do?

File: js/utils/ajax.js [hunk 4]
```diff
@@ -19,16 +19,30 @@
...
-  'jquery'
-], function($) {
+  'jquery',
+  'token-utils',
+  'user-utils'
+], function($, token, user) {
   $(document).ajaxError(function(event, jqueryXHR, error, errorThrown) {
+    var i, cluster,
+      isCurrentCluster = window.cluster && error.url.indexOf(window.cluster.api.url) > -1;
+
```
Comment (rezib) [inline js/utils/ajax.js]: The whole logic of this error handler is not really obvious. A large comment must be added here to explain to whole logic and all cases that should be handled and how.

File: js/utils/ajax.js (hunk 2 above)
Comment (Lorel) [inline js/utils/ajax.js]: find cluster concerned by error

**Comment by rezib** [comment]
I merged 002c79a, thanks for it!

File: js/core/index.js [hunk 5]
```diff
@@ -167,6 +167,9 @@ require([
       .success(function(data) {
         config.cluster.infos = data;
         $(document).trigger('show', { page: config.STARTPAGE });
+      })
```
Comment (rezib) [inline js/core/index.js]: I'm not sure what this change is supposed to fix, maybe the commit message would deserve few more sentences to make it clearer.

Also, I'm not sure the users should be redirected on the login page for every errors. There are several reasons for the /cluster to return error, including 500. The login form is relevant only in case of 403 and auth enabled on the cluster. In all other cases, it is pointless.

File: js/core/clusters/clusters.js [hunk 6]
```diff
@@ -53,6 +53,11 @@ define([
   return function(config) {
     var lock = false;
 
+    // getter for config.cluster
+    window.cluster = function() {
```
Comment (rezib) [inline js/core/clusters/clusters.js]: This is better than before but I still consider co-existence of window.cluster() and config.cluster as a technical debt. It must be fixed for the long term.

If we **need** to have window.cluster[()], then config.cluster must be removed.

File: js/core/clusters/clusters.js [hunk 7]
```diff
@@ -140,7 +162,7 @@ define([
         bindClusterButtons();
 
         // select first cluster once clusters informations have been retrieved
-        $(document).trigger('selectCluster', { clusterId: 0, $cluster: $('.cluster').first() });
+        return $(document).trigger('selectCluster', { clusterId: 0, $cluster: $('.cluster').first() });
```
Comment (rezib) [inline js/core/clusters/clusters.js]: Is this change related to the other changes of this commit? If not, this should be a separated commit.

File: js/utils/ajax.js [hunk 8]
```diff
@@ -19,16 +19,43 @@
...
+  }
+
   $(document).ajaxError(function(event, jqueryXHR, error, errorThrown) {
+    var i, cluster,
+      currentCluster = window.cluster(),
+      isCurrentCluster = currentCluster && error.url.indexOf(currentCluster.api.url) > -1;
+
+    // find cluster concerned by current error
+    for (i in window.clusters) {
+      cluster = window.clusters[i];
```
Comment (rezib) [inline js/utils/ajax.js]: Can you please rename the variable s/cluster/concernedCluster/?

File: js/utils/ajax.js [hunk 9]
```diff
@@ -19,16 +19,43 @@
...
+      if (cluster.api.url === extractDomain(error.url)) {
+        break;
+      }
+    }
+
     if (!jqueryXHR.status && !(error.url.indexOf('/authentication') > -1)) {
-      console.log(JSON.stringify(event), JSON.stringify(jqueryXHR), JSON.stringify(error), JSON.stringify(errorThrown));  // eslint-disable-line no-console
-      $('#flash .alert').append($('<p>').text('Error : unable to perform authentication'));
-      $('#flash').show();
+      // case when error status is not 403 (FORBIDDEN)
```
Comment (rezib) [inline js/utils/ajax.js]: The comment might be wrong since there is no check for != 403 here. Can you please explain in comment in what case the expression (!jqueryXHR.status && !(error.url.indexOf('/authentication') > -1)) is true.

File: js/core/clusters/clusters.js (hunk 6 above)
Comment (Lorel) [inline js/core/clusters/clusters.js]: `config` is not accessible for security reason, that's why a getter for `config.cluster` makes sense, no technical debt here

Comment (rezib) [inline js/core/clusters/clusters.js]: I did not say that config should removed, needless to convince me about its existence. I'm just talking about config.cluster here.

As soon as there are 2 ways to access the same information and it introduces confusion into other developers mind, it's a technical debt.

File: js/core/index.js (hunk 5 above)
Comment (Lorel) [inline js/core/index.js]: it fixes bug on login page refresh when accessing on another cluster where user is not yet authenticated: this page is not updated otherwise

File: js/core/clusters/clusters.js (hunk 6 above)
Comment (Lorel) [inline js/core/clusters/clusters.js]: there is only one way to set it, there is only one way to get it globally, when `config` is not is the current scope

**Comment by rezib** [comment]
No, we cannot fix #127 with bedfdd4 unfortunately. The shitty data come from pyslurm, the fix must be done up there. We also have to think about API clients != of the JS dashboard, those one also deserve correctly formatted data.

**Comment by Lorel** [comment]
@rezib, so you are meaning that `['swibfstl[46', '49-51', '53-54]']` is not a correct nodeset while `['swibfstl[46-47', '49-51', '53-54]']` would be a correct one ? I am not an expert about slurm, as you know, but here nothing is surprising me, and bedfdd401492f05e036b7f44c3a3cb5143325c62 handle this case

And I don't understand what you mean about crossdomain case and data

**Comment by rezib** [comment]
Both are wrong. PySLURM should return `['swibfstl[46,49-51,53-54]']` and `['swibfstl[46-47,49-51,53-54]']` without the intermediate quotes.

> And I don't understand what you mean about crossdomain case and data

My statement wasn't related to crossdomain but there are other REST API consumers than the JS dashboard. It is better if the REST API return valid data, to avoid implementing that kind of workaround in all API consumers/clients code bases.

**Comment by Lorel** [comment]
if you say so, but one year ago, `['swibfstl[46-47', '49-51', '53-54]']` was a correct data, and that's why "Topology" have been implemented based on this spec
so bedfdd401492f05e036b7f44c3a3cb5143325c62 is a fix to handle both `['swibfstl[46', '49-51', '53-54]']`, `['swibfstl[46-47', '49', '53-54]']` and `['swibfstl[46-47', '49-51', '54]']` formats
so it would be strange to reject it

and know that to send ['swibfstl[46-47,49-51,53-54]'] instead of `['swibfstl[46-47', '49-51', '53-54]']` will break the current "Topology" page implementation

**Comment by rezib** [comment]
Then we'll fix the topology view to eat valid data.
